### DISCUSSION TITLE
Pagination with getStaticProps

### DISCUSSION THREAD

**Discussion by mjr** [vercel/next.js#10987] [Help]
Would it be possible to do something very similar to getStaticProps (or using getStaticProps itself), but with pagination in props?

@Timer @timneutkens

**Comment by timneutkens** [comment]
Something like this should work, written from the top of my mind so haven't ran the code yet.

```
// pages/category/[category]/[page].js
import Link from 'next/link'
import {useRouter} from 'next/router'
import DefaultErrorPage from 'next/error'

export async function getStaticProps({ params }) {
  // fetch data based on category and page
  const {products, totalPages} = await getProductsByCategory(params.category, params.page)
  return {
    props: {
      products,
      category: params.category
      currentPage: params.page,
      totalPages
    }
  }
}

export async function getStaticPaths() {
  return {
    // Opt-in to on-demand generation for non-existent pages
    fallback: true,
    paths: [
       // return the paths needed here, this is just an example
       { params: { category: 'books', page: '1' } },
       { params: { category: 'books', page: '2' } },
       { params: { category: 'books', page: '3' } }
    ]
  }
}

export default function MyPage({products, currentPage, category, totalPages}) {
  const router = useRouter()

  // Show a loading state when the fallback is rendered
  if(router.isFallback) {
     return <h1>Loading...</h1>
  }

  // This includes setting the noindex header because static files always return a status 200 but the rendered not found page page should obviously not be indexed
  if(products.length === 0) {
    return <>
      <Head>
        <meta name="robots" content="noindex">
      </Head>
      <DefaultErrorPage statusCode={404} />
    </>
  }

  
  return <>
    <ul>
      {products.map((product) => {
         return <li>
           <h3>{product.title}</h3>
           <p>{product.description}</p>
         </li>
      })}
    </ul>
    <ul>
	  {Array.from({length: totalPages}, (_, index) => {
         return <li className={index === currentPage ? 'active' : undefined}>
           <Link href="/category/[category]/[page]" as={`/category/${category}/${index}`}>
             <a>Go to page {index}</a>
           </Link>
         </li>
      })}
    </ul>
  </>
}
```

Partially based on my answer here: https://github.com/zeit/next.js/discussions/10960#discussioncomment-1201

**Reply by mjr** [reply]
Thanks for your reply @timneutkens, your example works very well, but it was not exactly what i expected. What i was looking for would be something that fits the following example: i have a list of posts on the blog's home page, but i don't want to load all the preview of posts. (However i want to allow to fetch more from a "load more" button.)

**Reply by sandys** [reply]
either "load more" or "next/previous" with page numbers - both usecases are important. there is already a pull request - https://github.com/zeit/next.js/pull/11071 . i wonder if that can be updated.

**Reply by ckeeney** [reply]
@timneutkens this answer is great, but it requires having `/category/[slug]/[pageNumber]` as a path.  Is it possible to accomplish this with `getStaticProps` but keep the path `/category/[slug]?page=[pageNumber]`? 

I know it's possible using `getServerSideProps` but I'd like to keep the static generation if possible.

**Reply by ggepenyan** [reply]
Hey @ckeeney, did you find a solution for that?

**Reply by ckeeney** [reply]
For now I use `getServerSideProps` on this page.

**Comment by sandys** [comment]
@timneutkens  im very interested in this as well - we are trying to create a /blog section in our nextjs code. the "blog" folder in the source code will have about a hundred subdirectories - each representing a blog post in markdown (with its images in the same folder).

We are unable to figure out how to create static pages for each blog post and a paginated list page for the main blog landing page. 
do note that just the /blog section is static - the rest of the site is SSR. 
is there an example for this that we can refer ?

**Reply by timneutkens** [reply]
See the code I posted above.

**Reply by trompx** [reply]
Hey @timneutkens, it works with single level category but how do you do once you add subcategories? As the subcategory slug would potentially conflicts with category pageNumber. For exemple: /category/cars/1  and /category/cars/electriccars/1 where here the subcategory electriccars could potentially be set to a number.
Being able to handle that with query parameters would solve this.

**Comment by sjcodebook** [comment]
I am getting this error when deploying to zeit now => 
```
[POST] /api/postsApi
11:00:13:67
Status:
500
Duration:
8.1ms
Memory Used:
76 MB
ID:
kxq8t-1585546213659-5c3393750f30
User Agent:
axios/0.19.2
{
  fields: [ 'title', 'date', 'slug', 'author', 'coverImage', 'excerpt' ],
  page: 1
}
2020-03-30T05:30:13.688Z	572075eb-4a7a-47de-be16-072a9f7005f7	ERROR	Error: ENOENT: no such file or directory, scandir '/zeit/1cc63678/allPosts'
    at Object.readdirSync (fs.js:871:3)
    at getPostSlugs (/var/task/.next/serverless/pages/api/postsApi.js:306:52)
    at module.exports.fZHd.__webpack_exports__.default (/var/task/.next/serverless/pages/api/postsApi.js:253:86)
    at apiResolver (/var/task/node_modules/next/dist/next-server/server/api-utils.js:48:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  errno: -2,
  syscall: 'scandir',
  code: 'ENOENT',
  path: '/zeit/1cc63678/allPosts'
}
```
Workaround I'm using:

```
# next.config.js
module.exports = {
  env: {
    PROJECT_DIRNAME: __dirname,
  },
}
```
and in the api definition where i need the path(allPosts folder contains all blogs in markdown format and it is located in project root )

```
import fs from 'fs'
import { join } from 'path'

const postsDirectory = join(process.env.PROJECT_DIRNAME, 'allPosts')
```
It is working perfectly on local development.

**Reply by sandys** [reply]
@timneutkens we are stuck in a similar place. could you please help ? how do we create pagination in a blog using only markdown and deploy to now.sh ?

**Comment by sandys** [comment]
We have been struggling with the same thing

**Reply by 4ortytwo** [reply]
How did you go about it, @sandys?

**Comment by sandys** [comment]
I don't have a solution yet, unfortunately

**Comment by bhatsudo** [comment]
I have the same question. Any solutions yet?

**Comment by condinoaljoseph** [comment]
Here is my solution to my problem but its just a simple prev and next pagination. [https://aljoseph.vercel.app/posts/javascript-prototype](https://aljoseph.vercel.app/posts/javascript-prototype)

**Comment by ckeeney** [comment]
Related to/duplicate of https://github.com/vercel/next.js/discussions/17269

**Comment by aliahsan07** [comment]
does anyone have a solution to this problem?

**Comment by devsfer1** [comment]
Same here :/

**Comment by ng0cminh** [comment]
does anyone have a solution to this problem?

**Comment by Elijah-trillionz** [comment]
This is a pretty solid suggestion.
[It recommends fetching with getStatisProps and SWR (or sth similar) combined](https://www.reddit.com/r/nextjs/comments/l45yl2/comment/gkmkon3/)
